Shared/configurations/ConfigurationController.cpp

```diff
@@ -64,26 +79,183 @@ void ConfigurationController::select(int index)
+{
+  // any download other than the default from Esri should have it's own DsaAppConfig.json
+  // file present. if it is not present, the default will be created when the app is
+  // relaunched with this configuration selected.
+  const auto dsaAppConfigFilePath = configurationDirectory.absoluteFilePath(DsaUtility::FILE_NAME_APP_CONFIG);
+  QFile dsaAppConfigFile{dsaAppConfigFilePath};
+  if (!dsaAppConfigFile.exists() || !dsaAppConfigFile.open(QIODevice::ReadOnly))
+    return false;
+
+  // parse the file as a json document
+  QJsonParseError parseError;
+  const QByteArray configBytes{dsaAppConfigFile.readAll()};
+  const auto configDoc = QJsonDocument::fromJson(configBytes, &parseError);
+  dsaAppConfigFile.close();
+  if (configDoc.isNull() || configDoc.isEmpty() || !configDoc.isObject())
+    return false;
+
+  // find the previous root data directory value that was configured
+  auto configObj = configDoc.object();
+  const auto rootDataDirectoryVar = configObj.value(AppConstants::PROPERTYNAME_ROOT_DATA_DIRECTORY);
+  if (rootDataDirectoryVar.isUndefined())
+    return false;
+
+  // strip off any trailing slashes if they exist
+  auto rootDataDirectoryCleaned = rootDataDirectoryVar.toString();
+  while (rootDataDirectoryCleaned.back() == '\\' || rootDataDirectoryCleaned.back() == '/')
+    rootDataDirectoryCleaned.chop(1);
+
+  // overwrite the app config file with the new json doc
+  if (!dsaAppConfigFile.open(QIODevice::WriteOnly | QIODevice::Truncate))
+    return false;
+
+  // re-write the entire document with the old root directory replaced by the extracted location
+  QString configString{configBytes};
+  const QString rootDataDirectoryNew{configurationDirectory.absolutePath()};
+  const QString configStringUpdated{configString.replace(rootDataDirectoryCleaned, rootDataDirectoryNew)};
+  dsaAppConfigFile.write(configStringUpdated.toUtf8());
+  return true;
+}
+
+void ConfigurationController::resetConfigurationDeviceStatus(const QString& configurationName)
+{
+  if (!m_configurationListModel->setDataByName(configurationName, 0, ConfigurationListModel::PercentDownloaded) ||
+      !m_configurationListModel->setDataByName(configurationName, 0, ConfigurationListModel::PercentExtracted))
+    emit toolErrorOccurred(QStringLiteral("Error resetting the configurations status on the device"), QStringLiteral("Configuration Error"));
+}
+
 void ConfigurationController::download(int index)
 {
-  // prevent calling the download function if something is already in progress
-  if (m_downloadInProgress)
+  // fetch a copy of the selected model item from the listview
+  const auto configuration = m_configurationListModel->at(index);
+  const auto configurationUrl = configuration.url();
+  const auto configurationName = configuration.name();
+  if (configurationUrl.isLocalFile())
+  {
+    m_configurationListModel->setDataByName(configurationName, 100, ConfigurationListModel::PercentDownloaded);
```

Comment:
Just a passing comment, but I don't know what '100' means here.

Reply:
Maybe a templated setter like this makes it a little more readable?
```suggestion
    m_configurationListModel->set<ConfigurationListModel::PercentDownloaded>(configurationName, 100);
```

Reply:
That's a bit better, but I understand the meaning now after you've explained it. I think it's ok as-is, but primarily because that template stuff looks like overkill.